### PR TITLE
TestResult.init() is now a private method

### LINKED ISSUE #1893: Streamline invocation of "init" method within TestResult to be private
Currently there's a mix and match between how the `org.testng.internal.TestResult#init`method is invoked. Ideally speaking it should be confined only to `private` usage, but its not the case currently.

For more details please refer to this https://github.com/cbeust/testng/pull/1879#discussion_r208848031

This issue is to ensure that this discrepancy is fixed in TestNG before 7.0.0 is released.

### PR DESCRIPTION
**Pull request by krmahadevan**
Closes #1893

Fixes #1893  .

### Did you remember to?

- [X] Add test case(s)
- [X] Update `CHANGES.txt`

We encourage pull requests that:

* Add new features to TestNG (or)
* Fix bugs in TestNG

If your pull request involves fixing SonarQube issues then we would suggest that you please discuss this with the 
[TestNG-dev](https://groups.google.com/forum/#!forum/testng-dev) before you spend time working on it.

### CODE REVIEW

File: src/test/java/test/inject/InjectBeforeAndAfterMethodsWithTestResultSampleTest.java [hunk 1]
```diff
@@ -43,7 +45,9 @@ public void before(Method m, ITestResult r) {
   @AfterMethod
   public void after(Method m, ITestResult r) {
     String name = m.getName();
-    Assert.assertEquals(r, m_testResult);
+    //Dont run assertions based on built in "hashCode" comparison because its no longer the same
+    //object, but its a new ITestResult object but with the "Same" content.
```
Comment (juherr) [inline src/test/java/test/inject/InjectBeforeAndAfterMethodsWithTestResultSampleTest.java]: I'm not sure about that. Instead, TestResult should implement hashcode and equals methods.

**Comment by krmahadevan** [comment]
@juherr - I have rebased and fixed the review comment. Please help take a look.

**Comment by krmahadevan** [comment]
ping @juherr

**Review by juherr — APPROVED** [review]

**Comment by krmahadevan** [comment]
Based on an email from @cbeust am merging this PR, since @juherr has approved it.

**Pull request merged**